File: sidecar/kibrary_sidecar/protocol.py
from typing import Any, Literal, Optional
from pydantic import BaseModel

class Request(BaseModel):
    id: int
    method: str
    params: dict[str, Any] = {}

class ErrorBody(BaseModel):
    code: str
    message: str

class Response(BaseModel):
    id: int
    ok: bool
    result: Optional[Any] = None
    error: Optional[ErrorBody] = None

class Notification(BaseModel):
    event: str
    params: dict[str, Any] = {}

class ParsedRow(BaseModel):
    lcsc: str
    qty: int
    ok: bool
    error: Optional[str] = None

class ParseResult(BaseModel):
    rows: list[ParsedRow]
    format: Literal["bom", "list"]

File: src-tauri/src/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Serialize, Debug, Clone)]
pub struct Request {
    pub id: u64,
    pub method: String,
    pub params: Value,
}

#[derive(Deserialize, Debug, Clone)]
pub struct ErrorBody {
    pub code: String,
    pub message: String,
}

#[derive(Deserialize, Debug, Clone)]
pub struct Response {
    pub id: u64,
    pub ok: bool,
    #[serde(default)]
    pub result: Option<Value>,
    #[serde(default)]
    pub error: Option<ErrorBody>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct Notification {
    pub event: String,
    #[serde(default)]
    pub params: Value,
}
